TailLight/device.cpp

```diff
@@ -1,60 +1,119 @@
-    // Initialize tail-light to black to have control over HW state
-    WDF_TIMER_CONFIG timerCfg = {};
-    WDF_TIMER_CONFIG_INIT(&timerCfg, EvtSetBlackTimer);
 
-    WDF_OBJECT_ATTRIBUTES attribs = {};
-    WDF_OBJECT_ATTRIBUTES_INIT(&attribs);
-    attribs.ParentObject = device;
-    attribs.ExecutionLevel = WdfExecutionLevelPassive; // required to access HID functions
+    WDFDRIVER driver = WdfDeviceGetDriver(device);
+    DRIVER_CONTEXT* driverContext = WdfObjectGet_DRIVER_CONTEXT(driver);
 
-    WDFTIMER timer = nullptr;
-    NTSTATUS status = WdfTimerCreate(&timerCfg, &attribs, &timer);
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("WdfTimerCreate failed 0x%x\n", status));
-        return status;
-    }
+    //TRACE_FN_ENTRY;
 
-    status = WdfTimerStart(timer, 0); // no wait
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("WdfTimerStart failed 0x%x\n", status));
-        return status;
-    }
+    NTSTATUS status = STATUS_SUCCESS;
+
+    status = IoRegisterPlugPlayNotification(
+        EventCategoryDeviceInterfaceChange,
+        PNPNOTIFY_DEVICE_INTERFACE_INCLUDE_EXISTING_INTERFACES,
+        (PVOID)&GUID_DEVINTERFACE_HID,
+        WdfDriverWdmGetDriverObject(driver),
+        PnpNotifyDeviceInterfaceChange,
+        (PVOID)device,
+        &driverContext->pnpDevInterfaceChangedHandle
+    );
+
+    //TRACE_FN_EXIT;
 
     return status;
 }
 
 
-static UNICODE_STRING GetDevicePropertyString(WDFDEVICE device, DEVICE_REGISTRY_PROPERTY DeviceProperty) {
+
+static NTSTATUS WdfDeviceAllocAndQueryPropertyFrom(WDFOBJECT Source, 
+    DEVICE_REGISTRY_PROPERTY DeviceProperty, 
+    PWDF_OBJECT_ATTRIBUTES attributes, 
+    WDFMEMORY* memory) {
+    return WdfDeviceAllocAndQueryProperty((WDFDEVICE)Source, 
+        DeviceProperty, 
+        NonPagedPoolNx, 
+        attributes, 
+        memory);
+}
+
+NTSTATUS WdfIoTargetAllocAndQueryTargetPropertyFrom(WDFOBJECT Source, 
+    DEVICE_REGISTRY_PROPERTY DeviceProperty, 
+    PWDF_OBJECT_ATTRIBUTES attributes, 
+    WDFMEMORY* memory) {
+    return WdfIoTargetAllocAndQueryTargetProperty((WDFIOTARGET)Source, 
+        DeviceProperty, 
+        NonPagedPoolNx, 
+        attributes, 
+        memory);
+}
+
+UNICODE_STRING GetDevicePropertyString(WDFOBJECT Source,
+    DEVICE_REGISTRY_PROPERTY DeviceProperty, 
+    PFN_WDFALLOCANDQUERYTARGETPROPERTYFROM pfnWdfAllocAndQueryTargetPropertyFrom) {
+
+    NT_ASSERTMSG("pfnWdfAllocAndQueryTargetPropertyFrom invalid", 
+        (PULONG_PTR)pfnWdfAllocAndQueryTargetPropertyFrom > (PULONG_PTR)65535);
+    
     WDF_OBJECT_ATTRIBUTES attributes = {};
     WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
-    attributes.ParentObject = device; // auto-delete with device
+    attributes.ParentObject = (WDFOBJECT)Source; // auto-delete with device
 
     WDFMEMORY memory = 0;
-    NTSTATUS status = WdfDeviceAllocAndQueryProperty(device, DeviceProperty, NonPagedPoolNx, &attributes, &memory);
+    NTSTATUS status = pfnWdfAllocAndQueryTargetPropertyFrom(Source, 
```

Comment:
Wouldn't it be simpler to instead always call `WdfIoTargetAllocAndQueryTargetProperty` here, instead of using a function pointer?

The `WDFDEVICE`-based client's can then call `WdfDeviceGetIoTarget(device)` to get the local I/O target for the device, so that an `WDFIOTARGET` is always present.

Reply:
Alternative proposal that avoids having to introduce a function pointer: https://github.com/forderud/IntelliMouseDriver/pull/63